browser/module.go

```diff
@@ -69,3 +76,459 @@ func (*RootModule) NewModuleInstance(vu k6modules.VU) k6modules.Instance {
 func (mi *ModuleInstance) Exports() k6modules.Exports {
 	return k6modules.Exports{Default: mi.mod}
 }
+
+type mapping map[string]any
+
+// mapBrowserToGoja maps the browser API to the JS module.
+// The motivation of this mapping was to support $ and $$ wildcard
+// methods.
+// See issue #661 for more details.
+func mapBrowserToGoja(vu k6modules.VU) *goja.Object {
+	rt := vu.Runtime()
+
+	var (
+		obj         = rt.NewObject()
+		browserType = chromium.NewBrowserType(vu)
+	)
+	for k, v := range mapBrowserType(rt, browserType) {
+		err := obj.Set(k, rt.ToValue(v))
+		if err != nil {
+			k6common.Throw(rt, fmt.Errorf("mapping: %w", err))
+		}
+	}
+
+	return obj
+}
+
+func mapRequest(rt *goja.Runtime, r api.Request) mapping {
```

Comment:
Are we having to `mapRequest` and `mapResponse` because it can return a `frame`, which in turn can return `pages`? Seems like this is going to catch people out as we will forget to add new methods to the mappings. I preferred the approach where we looped through all the methods and just set them to the object, like above in `mapBrowserToGoja`.

Reply:
Hi there!

I agree that the solution we have in place is not future-proof, but it's good enough for now. In the future, we should consider automating or generating this mapping code.

I've added some tests to verify the behavior, which you can find in the following commits:

* https://github.com/grafana/xk6-browser/pull/673/commits/d062048411cf225ca00da0b9f206b148478e9a8f
* https://github.com/grafana/xk6-browser/pull/673/commits/0c388d50abb43a43f1ab0c6c88e6f274839b47f6

On the bright side, if something goes wrong and a method can't be found, users will receive an error message. We can fix any issues that arise, and with the tests in place, the likelihood of that happening is low.

Additionally, these manual mappings will allow us to move the goja layer from our domain logic code, making it more type-safe and easier to work with. I'll create an issue about that soon.

Let me know if you have any questions or concerns.

---

Related:

[The PR description](https://github.com/grafana/xk6-browser/pull/673#issue-1494144809):

> As discussed in the issue https://github.com/grafana/xk6-browser/issues/661, we're now "manually" mapping the browser module methods to Goja. The global Goja field name mappers are removed from the code. This can be automated in the future as the mapping can be error-prone.

Also [here](https://github.com/grafana/xk6-browser/issues/661):

> However, not sure the current solution will be future-proof. Since we have a relatively large API surface area; it can be error-prone to manually do this kind of mappings. I believe we might want to automate this in the future (maybe by writing a source-code generator). That way the resulting mapping can be in the code itself and people can easily track them without magic.

Reply:
Replied [here](https://github.com/grafana/xk6-browser/pull/673#issuecomment-1350594447).